src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -1761,7 +1777,9 @@ pub fn is_useful<'p, 'tcx>(
             // `(<direction-1>, <direction-2>, true)` - we are
             // satisfied with `(_, _, true)`. In this case,
             // `used_ctors` is empty.
-            if missing_ctors.all_ctors_are_missing() {
+            // The exception is: if we are at the top-level, for example in an empty match, we
+            // prefer reporting the list of constructors instead of just `_`.
+            if missing_ctors.all_ctors_are_missing() && !is_top_level {
```

Comment:
We should special case integer ranges here, as `u8::MIN…=u8::MAX`, etc. isn't any more helpful. I'm happy with the overall change though.